### ISSUE TITLE
MV updates are issued with an infinite timeout - a recipe for a live-lock

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#10031]
*Installation details*
HEAD: fe65122ccd40a2a3577121aebdb9a5b50deb4a90

*Description*
Materialized views update (the write part) is sending mutations with an infinite timeout.
This is very dangerous despite the fact that there is a MV Flow Control (MV FC) at place which is going to delay a response to a corresponding base table update.

Imagine that a write to a base table times out because of a temporary cluster overload which is caused by MVs write being heavy and slow and client retries before the corresponding MV mutation ends. 

As a result a new MV is going to be queued and MV FC is going to make sure this base table update times out too but there are already 2 queued MV mutations and so on. 

As a result we get a live-lock.

A simple solution could be to use the same timeout for MV updates as we use for the corresponding base table update. We already pass it to the flat mutation reader in the same flow:

* https://github.com/scylladb/scylla/blob/next-5.0/db/view/view.cc#L1296
* https://github.com/scylladb/scylla/blob/next-5.0/replica/table.cc#L1750

**Comment by vladzcloudius** [comment]
@nyh @psarna FYI

**Comment by avikivity** [comment]
A view build consists of numerous steps. If each step has a timeout, then the small probability of a timeout in each step multiplied by the total number of build steps results in low probability of the build succeeding.

I don't understand the live lock scenario. The builder is going through a finite (but large) number of build steps. The concurrency is bounded. This happens in a separate scheduling group so while there can be impact on regular writes, it is bounded.

**Comment by nyh** [comment]
> _Description_ Materialized views update (the write part) is sending mutations with an infinite timeout.

It's not **infinite**, but it is indeed large:

While there is no timeout for the _read_ part (scanning the base-table data), the write part does have a timeout. The view building code calls `table::populate_views()` which calls `db::view::mutate_MV()` and that ends up calling `send_to_endpoint()` with `db::write_type::VIEW` - and that uses a 5min timeout in that case (it's the same as the regular view updates - I couldn't find a separate code path). Not an infinite timeout.

Arguably, 5 minutes is still a huge timeout - maybe that is what you called infinite. I seem to recall an argument between Duarte, Avi, and perhaps (?) Gleb, one what this timeout should be. 5 minutes was a compromise - Duarte wanted it to be longer (see commit e33e1870963b45db3dab37e81fdce47df43a2e55).

> This is very dangerous despite the fact that there is a MV Flow Control (MV FC) at place which is going to delay a response to a corresponding base table update.

As Avi noted, the **view building** code (kicking in when adding a view to an existing table) has a very low concurrency, so it can't possibly trigger the view-update flow control, which only kicks in when there is a large backlog of view updates. At most the view building will add a fixed size (I believe 1MB, see commit fec562ec8ff0b0ccaf9f72b8feb23fdc6826327d) to the backlog.
 
> Imagine that a write to a base table times out because of a temporary cluster overload which is caused by MVs write being heavy and slow and client retries before the corresponding MV mutation ends.

A write to the base table won't wait for the view updates to complete. What *can* happen is that because we have a huge number of unfinished view updates we start slowing down the base-table response deliberately - this is the MV flow control. If we reach a point that *this* times out, we must have very high client concurrency, otherwise the new requests would have slowed down so much that the old ones would have time to finish. You're right that for a given client timeout there will indeed be some (very large) concurrency that will cause the average (or some percentile) latency to be higher than the client timeout - and we end up with zero performance. I think @psarna once looked into that - and into load shedding to fix it. But this will only happen if we have very large client concurrency - and I don't see how it's related to view building.

By the way, unrelated to this issue - the mere fact that we already have a huge number of unfinished view writes causes its own problems - see  #8873 and  #4940.
 
> As a result a new MV is going to be queued and MV FC is going to make sure this base table update times out too but there are already 2 queued MV mutations and so on.
> 
> As a result we get a live-lock.
> 
> A simple solution could be to use the same timeout for MV updates as we use for the corresponding base table update.

We went in the opposite direction (see commit e33e1870963b45db3dab37e81fdce47df43a2e55). A basic design decision (arguably, a regrettable decision) of Cassandra's and Scylla's MV is that view updates are usually remote and asynchronous (the client does **not** wait for them). Because we never wait for view writes to finish, we don't care if they finish in 1 second or 10 seconds, even if the client wants a respond by 1 second. This is we ended up with allowing huge view-update backlogs (even 5 minute worth) and only starting flow control only when they are huge (the controller function is gives a tiny delay unless a big percentage of the backlog queue is full). I agree that those typically-large backlogs are problematic and I was actually against them. #4940 discusses other problems in how these backlogs are handled.

I think that instead of theorizing on possible problems, we should try to find specific cases where these algorithms break and then fix those cases. We tried to do this once, but 1. it wasn't easy to find breaking examples, and 2. we always stopped this work in the middle and moved to something else, so it didn't result in anything.

> We already pass it to the flat mutation reader in the same flow:
> 
> * https://github.com/scylladb/scylla/blob/next-5.0/db/view/view.cc#L1296
> * https://github.com/scylladb/scylla/blob/next-5.0/replica/table.cc#L1750

I didn't understand what you are saying here. And this is the *reader*, not the writer.

**Comment by vladzcloudius** [comment]
> A view build consists of numerous steps. If each step has a timeout, then the small probability of a timeout in each step multiplied by the total number of build steps results in low probability of the build succeeding.
> 
> I don't understand the live lock scenario. The builder is going through a finite (but large) number of build steps. The concurrency is bounded. This happens in a separate scheduling group so while there can be impact on regular writes, it is bounded.

@avikivity The issue is not about view *building* stage - but rather about view *update*.
It's the sage when view is built and there is an incoming write to a base table that triggers an update to a view.

**Comment by vladzcloudius** [comment]
>>  We already pass it to the flat mutation reader in the same flow:
>>
>>  https://github.com/scylladb/scylla/blob/next-5.0/db/view/view.cc#L1296
>>  https://github.com/scylladb/scylla/blob/next-5.0/replica/table.cc#L1750

> I didn't understand what you are saying here. And this is the reader, not the writer.

@nyh 
I wrote that we already pass a base table write timeout to a read part of a view update - so why not to pass it to the write part too which today uses an infinite timeout (`clock_type::time_point::max()` to be specific): see a default value of a `timeout` parameter of a corresponding `mutate_local()` call.

**Comment by psarna** [comment]
@vladzcloudius the link you referred to is a local view update (as the variable name denotes), which doesn't send anything over the network, but instead applies everything straight to this node's storage.

(interlude: It might be subject to a separate debate whether we want an infinite timeout in that case, but maybe we do - a local disk is likely to respond in finite time, and if it doesn't, the whole node is screwed anyway.)

Sending view updates to remote nodes happens via storage proxy's `send_to_endpoint`, which applies a 5 minute timeout for view updates, as Nadav said. Ref:
https://github.com/scylladb/scylla/blob/00a6fda7b98438184024fc4683b0accf1852e30c/service/storage_proxy.cc#L2591-L2607

**Comment by avikivity** [comment]
A timeout forks reality into two versions, one that timed out and one that didn't. To prevent an increase in the number of realities and growth of the test matrix, the number of timeout gated operations should be kept to a minimum.

Timeouts are useful in two cases: where there is a user deadline, and in order to prevent failure of a remote node from killing the local node. We don't have a user deadline here (the operation is asynchronous wrt. the user write), and we have a separate mechanism noticing that the remote node is down and failing the operation (IIRC). So I think an infinite timeout is correct here.

**Comment by kostja** [comment]
I agree with Avi the timeout is correct, however, I think the underlying pain here is that we have no single view update queue per receiving peer. 

I'd suggest to change the view update logic to write the hint first, and then send the hints in a background worker. This is currently somewhat the case, but the difference is that today each view update generates an own background future, which potentially creates unlimited concurrency, so needs to be controlled by a semaphore. Writing the hint could happen in foreground, which will bound the concurrency, whereas the hint propagation is already taking place with limited concurrency. 

Anyhow, there is more than one way to skin the cat here, I think it's recognized that view update concurrency needs more work, even after @Jadw1 patches. I don't see a point in keeping this specific ticket open, we will be specifying the next steps out in a design anyway.

**Comment by nyh** [comment]
@kostja in general, I think it's a stategic mistake to close issues about existing bugs, just because those bugs will not exist in some specific future re-implementation you had in mind for the code - a re-implementation that there is no promise we'll actually do, or any idea when we'll do it.

However, this specific issue I never really understood what it is about, or what kind of reproducer might reproduce it, so I'm fine with closing this issue. CC @vladzcloudius who originally reported it and may not agree with this decision.

**Comment by vladzcloudius** [comment]
> However, this specific issue I never really understood what it is about, or what kind of reproducer might reproduce it, so I'm fine with closing this issue. CC @vladzcloudius who originally reported it and may not agree with this decision.

@nyh @kostja @avikivity after re-reading the Piort Sarna's comment that the infinite timeout is actually only on the local view update path the reproducer and the seriousness of the bug is even clearer: 
* Local view updates are synchronous (there won't be any views in the picture in general)
1) If the read part of the view update which uses an infinite timoeut takes more than the the (base table) write timeout that (base table) write is going to be deemed failed however the corresponding read continuation would survive.
2) The corresponding CQL base table write may timeout as well and the following (base table) write retry will create another everlasting read continuation.

The repeated sequence of (1) and (2) can create an infinite amount of such read continuation which in line can cause OOM in the best case and live-lock in the worst.

Please, let me know if I missed anything.

And, yes, @kostja, I agree with @nyh - let's not close issues unless they are fixed.
